internal/resolve/catalog_test.go

```diff
@@ -963,3 +963,150 @@ func TestMultipleChannels(t *testing.T) {
 	assert.Equal(t, bsemver.MustParse("2.0.0"), *gotVersion)
 	assert.Equal(t, ptr.To(packageDeprecation(pkgName)), gotDeprecation)
 }
+
+func TestAllCatalogsDisabled(t *testing.T) {
+	pkgName := randPkg()
+	listCatalogs := func(ctx context.Context, options ...client.ListOption) ([]catalogd.ClusterCatalog, error) {
+		return []catalogd.ClusterCatalog{
+			{
+				Spec: catalogd.ClusterCatalogSpec{
+					Availability: "Disabled",
+				},
+			},
+			{
+				Spec: catalogd.ClusterCatalogSpec{
+					Availability: "Disabled",
+				},
+			},
+		}, nil
+	}
+
+	getPackage := func(ctx context.Context, cat *catalogd.ClusterCatalog, packageName string) (*declcfg.DeclarativeConfig, error) {
+		return genPackage(pkgName), nil
+	}
+
+	r := CatalogResolver{
+		WalkCatalogsFunc: CatalogWalker(listCatalogs, getPackage),
+	}
+
+	ce := buildFooClusterExtension(pkgName, []string{}, ">=1.0.0", ocv1alpha1.UpgradeConstraintPolicyCatalogProvided)
+	_, _, _, err := r.Resolve(context.Background(), ce, nil)
+	require.Error(t, err)
+	assert.Contains(t, err.Error(), "no enabled catalogs found for package")
+}
+
+func TestSomeCatalogsDisabled(t *testing.T) {
+	pkgName := randPkg()
+	listCatalogs := func(ctx context.Context, options ...client.ListOption) ([]catalogd.ClusterCatalog, error) {
+		return []catalogd.ClusterCatalog{
+			{
+				Spec: catalogd.ClusterCatalogSpec{
+					Availability: "Enabled",
+				},
+			},
+			{
+				Spec: catalogd.ClusterCatalogSpec{
+					Availability: "Disabled",
+				},
+			},
+		}, nil
+	}
+
+	getPackage := func(ctx context.Context, cat *catalogd.ClusterCatalog, packageName string) (*declcfg.DeclarativeConfig, error) {
+		return genPackage(pkgName), nil
```

Comment:
In order to properly test that we ignore disabled catalogs, don't we need to the enabled and disabled catalogs to have disjoint sets of bundle versions, with the disable catalog having higher versions than the enabled catalog? That way if we ever somehow stop ignoring disabled catalogs, resolution would incorrectly resolve to a bundle we find in the disabled catalog and this test would fail.

Reply:
Yes, we've modified the `getPackage` function to return a higher version for the disabled catalog and a lower version for the enabled catalog. This ensures that if the disabled catalog is incorrectly processed, the test will fail by selecting the higher version from the disabled catalog.